File: src/Tarefa.hs

module Tarefa where
import Usuario (Usuario(..))


data StatusTarefa = Backlog | Pendente | EmDesenvolvimento | Concluido
  deriving (Eq)

instance Show StatusTarefa where
  show Backlog = "Backlog"
  show Pendente = "Pendente"
  show EmDesenvolvimento = "Em Desenvolvimento"
  show Concluido = "Concluído"
data Tarefa = Tarefa
  { tarefaId :: Int
  , tarefaTitulo :: String
  , tarefaDescricao :: String
  , tarefaPrioridade :: Int
  , tarefaStatus :: StatusTarefa
  , tarefaIdCriador :: Int
  , tarefaIdResponsavel :: Int
  , tarefaEmpresaId :: Int
  } deriving (Show, Eq)

backlogEmpresa :: Usuario -> [Tarefa] -> IO [Tarefa]
backlogEmpresa usuario tarefas = do
  let tarefasBacklog = filter (\t -> tarefaStatus t == Backlog) tarefas
  putStrLn "\nBacklog de Tarefas da Empresa:"
  mapM_ (\t -> putStrLn $ show (tarefaId t) ++ ": " ++ tarefaTitulo t ++ " - " ++ show (tarefaStatus t)) tarefasBacklog
  putStrLn "\nOpções:"
  putStrLn "1. Criar nova tarefa"
  putStrLn "2. Voltar ao menu principal"
  putStrLn "Escolha uma opção: "
  opcao <- getLine
  case opcao of
    "1" -> do
      novaTarefa <- adicionarTarefa usuario
      backlogEmpresa usuario (novaTarefa : tarefas)  
    "2" -> return tarefas  
    _   -> do
      putStrLn "Opção inválida, tente novamente."
      backlogEmpresa usuario tarefas  

adicionarTarefa :: Usuario -> IO Tarefa
adicionarTarefa usuario = do
  putStrLn "Digite o ID da Tarefa:"
  id <- readLn
  putStrLn "Digite o Título da Tarefa:"
  titulo <- getLine
  putStrLn "Digite a Descrição da Tarefa:"
  descricao <- getLine
  putStrLn "Digite a Prioridade da Tarefa (1-5):"
  prioridade <- readLn
  let empresaId = usuarioEmpresaId usuario 
  let criadorId = usuarioId usuario 
  let tarefa = Tarefa id titulo descricao prioridade Backlog criadorId 0 empresaId
  putStrLn $ "Tarefa adicionada: " ++ show tarefa
  return tarefa

-- Função para listar as tarefas do usuário e permitir a modificação de status
listarTarefasUsuario :: Usuario -> [Tarefa] -> IO [Tarefa]
listarTarefasUsuario usuario tarefas = do
  let tarefasUsuario = filter (\t -> tarefaIdCriador t == usuarioId usuario || tarefaIdResponsavel t == usuarioId usuario) tarefas
  if null tarefasUsuario
    then do
      putStrLn "O usuário não possui tarefas."
      return tarefas -- Retorna a lista sem modificações
    else do
      putStrLn "Tarefas do usuário:"
      mapM_ (\t -> putStrLn $ "ID: " ++ show (tarefaId t) ++ ", Título: " ++ tarefaTitulo t ++ ", Status: " ++ show (tarefaStatus t)) tarefasUsuario
      putStrLn "Digite o ID da tarefa que deseja modificar o status ou 0 para sair:"
      tarefaIdStr <- getLine
      let tarefaIdEscolhida = read tarefaIdStr :: Int
      if tarefaIdEscolhida == 0
        then return tarefas -- Sai sem modificar
        else do
          -- Verifica se a tarefa existe
          let tarefaSelecionada = filter (\t -> tarefaId t == tarefaIdEscolhida) tarefasUsuario
          if null tarefaSelecionada
            then do
              putStrLn "Tarefa não encontrada."
              listarTarefasUsuario usuario tarefas -- Chama a função novamente
            else do
              putStrLn "Digite o novo status da tarefa (2 para Pendente, 3 para EmDesenvolvimento, 4 para Concluido):"
              novoStatusEntrada <- getLine
              let novoStatus = case novoStatusEntrada of
                    "2" -> Pendente
                    "3" -> EmDesenvolvimento
                    "4" -> Concluido
                    _   -> tarefaStatus (head tarefaSelecionada) -- Se a entrada for inválida, mantém o status atual
              let tarefasAtualizadas = map (atualizarStatusTarefa tarefaIdEscolhida novoStatus) tarefas
              putStrLn "Status atualizado com sucesso!"
              return tarefasAtualizadas

-- Instância Show para StatusTarefa para facilitar a conversão para String


-- Função auxiliar para atualizar o status de uma tarefa
atualizarStatusTarefa :: Int -> StatusTarefa -> Tarefa -> Tarefa
atualizarStatusTarefa idTarefa novoStatus tarefa =
  if tarefaId tarefa == idTarefa
    then tarefa { tarefaStatus = novoStatus }
    else tarefa

File: prolog/src/tarefa.pl
:- module(tarefa, [adicionar_tarefa/1, listar_tarefas_backlog/1, listar_tarefas_usuario/1]).

:- use_module(usuario).

% Definição do status da tarefa
status_tarefa(backlog).
status_tarefa(pendente).
status_tarefa(em_desenvolvimento).
status_tarefa(concluido).

% Estrutura da tarefa
:- dynamic tarefa/7.


% Função para listar as tarefas com status de 'backlog' para uma empresa específica
listar_tarefas_backlog(Usuario) :- 
    usuario_empresa_id(Usuario, EmpresaId),  % Obtém o ID da empresa do usuário
    findall(tarefa(Id, Titulo, Descricao, Prioridade, backlog, CriadorId, ResponsavelId, EmpresaId), 
            tarefa(Id, Titulo, Descricao, Prioridade, backlog, CriadorId, ResponsavelId, EmpresaId), 
            Tarefas),
    (   Tarefas = []
    ->  writeln('Não há tarefas no backlog para esta empresa.')
    ;   listar_tarefas(Tarefas),
        writeln('\nOpções:'),
        writeln('1. Criar nova tarefa'),
        writeln('2. Voltar ao menu anterior'),
        read(Opcao),
        processar_opcao(Opcao, Usuario)
    ).

% Função para listar as tarefas
listar_tarefas([]).
listar_tarefas([Tarefa | Resto]) :- 
    imprimir_tarefa(Tarefa),
    listar_tarefas(Resto).

% Função para processar a opção escolhida
processar_opcao(1, Usuario) :- 
    adicionar_tarefa(Usuario),  % Chama a função para adicionar uma nova tarefa
    listar_tarefas_backlog(Usuario).  % Chama a função novamente para mostrar as tarefas atualizadas
processar_opcao(2, _) :- 
    writeln('Voltando ao menu anterior.').
processar_opcao(_, Usuario) :- 
    writeln('Opção inválida, tente novamente.'),
    listar_tarefas_backlog(Usuario).  % Chama a função novamente para mostrar as tarefas

% Função para adicionar uma nova tarefa
adicionar_tarefa(Usuario) :- 
    writeln('Digite o ID da Tarefa:'),
    read(Id),
    writeln('Digite o Título da Tarefa:'),
    read(Titulo),
    writeln('Digite a Descrição da Tarefa:'),
    read(Descricao),
    writeln('Digite a Prioridade da Tarefa (1-5):'),
    read(Prioridade),
    % Obtendo IDs da empresa e criador
    usuario_empresa_id(Usuario, EmpresaId),
    usuario_id(Usuario, CriadorId),
    % Criando a tarefa com status 'backlog'
    Tarefa = tarefa(Id, Titulo, Descricao, Prioridade, backlog, CriadorId, 0, EmpresaId),
    assertz(Tarefa),  % Adiciona a nova tarefa ao banco de dados dinâmico
    writeln('Tarefa adicionada com sucesso:'),
    imprimir_tarefa(Tarefa).

% Predicado para verificar se a tarefa pertence ao usuário (pelo CriadorId ou ResponsavelId)
tarefa_de_usuario(Usuario, tarefa(_, _, _, _, _, CriadorId, _, _)) :- 
    usuario_id(Usuario, CriadorId).
tarefa_de_usuario(Usuario, tarefa(_, _, _, _, _, _, ResponsavelId, _)) :- 
    usuario_id(Usuario, ResponsavelId).


% Função para listar as tarefas do usuário
listar_tarefas_usuario(Usuario) :- 
    findall(tarefa(Id, Titulo, Descricao, Prioridade, Status, CriadorId, ResponsavelId), 
            tarefa(Id, Titulo, Descricao, Prioridade, Status, CriadorId, ResponsavelId), 
            Tarefas),  % Obtém todas as tarefas
    include(tarefa_de_usuario(Usuario), Tarefas, TarefasUsuario),  % Filtra tarefas do usuário
    (   TarefasUsuario = []
    ->  write('O usuário não possui tarefas.'), nl
    ;   write('Tarefas do usuário:'), nl,
        maplist(imprimir_tarefa, TarefasUsuario),  % Imprime as tarefas do usuário
        write('Digite o ID da tarefa que deseja modificar o status ou 0 para sair:'), nl,
        read(TarefaIdEscolhida),
        (   TarefaIdEscolhida = 0
        ->  nl  % Apenas nova linha para melhor formatação
        ;   modificar_status(TarefaIdEscolhida)  % Chama a função para modificar o status
        )
    ).

% Função para imprimir uma tarefa
imprimir_tarefa(tarefa(Id, Titulo, Descricao, Prioridade, Status, CriadorId, ResponsavelId, _)) :- 
    format('ID: ~w, Título: ~w, Descrição: ~w, Prioridade: ~w, Status: ~w, Criador: ~w, Responsável: ~w~n', 
           [Id, Titulo, Descricao, Prioridade, Status, CriadorId, ResponsavelId]).

% Função para modificar o status da tarefa
modificar_status(TarefaId) :- 
    retract(tarefa(TarefaId, Titulo, Descricao, Prioridade, _, CriadorId, ResponsavelId, EmpresaId)),  % Remove a tarefa antiga
    writeln('Digite o novo status para a tarefa (backlog, pendente, em_desenvolvimento, concluido):'),
    read(NovoStatus),
    (   status_tarefa(NovoStatus) ->  % Verifica se o novo status é válido
        assertz(tarefa(TarefaId, Titulo, Descricao, Prioridade, NovoStatus, CriadorId, ResponsavelId, EmpresaId)),  % Adiciona a tarefa com novo status
        writeln('Status da tarefa modificado com sucesso.')
    ;   writeln('Status inválido. A tarefa não foi alterada.')
    ).
